File: backend/app/models/scan.py
from datetime import UTC, datetime
from enum import Enum
from uuid import UUID, uuid4

from pydantic import BaseModel, ConfigDict, Field


class ScanStatus(str, Enum):
    IDLE = "idle"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class ScanType(str, Enum):
    QUICK = "quick"  # ARP scan
    FULL = "full"  # ARP + Port scan
    PASSIVE = "passive"  # Sniffing only


class ScanRequest(BaseModel):
    """Request to start a network scan."""

    type: ScanType = Field(
        default=ScanType.QUICK, description="Type of scan to perform"
    )
    target_subnets: list[str] | None = Field(
        None,
        description=(
            "Specific subnets to scan (e.g. 192.168.1.0/24). If None, auto-detect."
        ),
    )

    model_config = ConfigDict(from_attributes=True)


class ScanResult(BaseModel):
    """Result of a scan operation."""

    id: UUID = Field(default_factory=uuid4, description="Unique scan ID")
    status: ScanStatus = Field(..., description="Final status of the scan")
    started_at: datetime = Field(
        default_factory=lambda: datetime.now(UTC), description="Start timestamp"
    )
    completed_at: datetime | None = Field(None, description="Completion timestamp")
    devices_found: int = Field(0, description="Number of devices discovered")
    error: str | None = Field(None, description="Error message if failed")

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/lib/services/scan.ts
import { api } from '../api';

export interface ScanResult {
  id: string;
  status: string;
  devices_found: number;
  started_at?: string;
  completed_at?: string | null;
  error?: string | null;
}

export interface ScanRequest {
  type?: 'quick' | 'full' | 'passive';
  target_subnets?: string[];
}

export const scanService = {
  startScan: async (request: ScanRequest = {}) => {
    // Default to quick scan if not specified
    const scanRequest = {
      type: request.type || 'quick',
      target_subnets: request.target_subnets || null,
    };
    const response = await api.post('/scan/start', scanRequest);
    return response.data;
  },

  getScanStatus: async (): Promise<ScanResult> => {
    const response = await api.get<ScanResult>('/scan/status');
    return response.data;
  },
};
